File: tlg/src/main/java/ch/bailu/tlg/StateInit.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.IOException;

public class StateInit extends State {
    public final static int MATRIX_WIDTH=10;
    public final static int MATRIX_HEIGHT=20;

	public StateInit(InternalContext c) {
		super(c);
	}


	
	public State init(PlatformContext c) {
		test(c);
		
	    try {
			return restore(c);
		} catch (IOException e) {
			return create(c);
		}
	    
		
	}
	
	public static void test(PlatformContext c) {
		//HighscoreList.test(c);
	}



	private State create(PlatformContext gc) {
	    context.previewMatrix  = new MatrixWithShape(5,5);
	    context.mainMatrix     = new MatrixLineManipulator(MATRIX_WIDTH,MATRIX_HEIGHT);
	    context.currentScore   = new Score();
	    context.textBox        = new TextBox();
	    
	    
	    State state = new StateRunning(context);
	    state.init(gc);
	    return state;
	}

	
	private State restore(PlatformContext gc) throws IOException {
		BufferedInputStream input = gc.getInputStream(InternalContext.STATE_FILE);
	
		int version = ByteInteger.read(input);
		if (version != InternalContext.STATE_FILE_VERSION) throw new IOException();


		context.previewMatrix = new MatrixWithShape(input);
		context.mainMatrix    = new MatrixLineManipulator(input);
		context.currentScore  = new Score(input);
		context.textBox       = new TextBox();
		
		State state = State.restoreContextFactory(context, ByteInteger.read(input));
		
		input.close();
		return state;
	}



	@Override
	public int getID() {
		return 0;
	}

}

File: tlg/src/main/kotlin/state/StateInit.kt
package state

import TlgConfiguration.MATRIX_HEIGHT
import TlgConfiguration.MATRIX_WIDTH
import TlgConfiguration.STATE_FILE
import TlgConfiguration.STATE_FILE_VERSION
import context.ByteInteger
import context.InternalContext
import context.PlatformContext
import matrix.MatrixLineManipulator
import matrix.MatrixWithShape
import score.Score
import java.io.IOException

class StateInit(c: InternalContext) : State(c) {
    override fun init(c: PlatformContext): State {
        return try {
            restore(c)
        } catch (e: IOException) {
            create(c)
        }
    }

    private fun create(gc: PlatformContext): State {
        context.previewMatrix = MatrixWithShape(5, 5)
        context.mainMatrix = MatrixLineManipulator(MATRIX_WIDTH, MATRIX_HEIGHT)
        context.currentScore = Score()
        val state: State = StateRunning(context)
        state.init(gc)
        return state
    }

    @Throws(IOException::class)
    private fun restore(gc: PlatformContext): State {
        val input = gc.getInputStream(STATE_FILE)
        val version = ByteInteger.read(input)
        if (version != STATE_FILE_VERSION) throw IOException()
        context.previewMatrix = MatrixWithShape(input)
        context.mainMatrix = MatrixLineManipulator(input)
        context.currentScore = Score(input)
        val state = restoreContextFactory(context, ByteInteger.read(input))
        input.close()
        return state
    }

    override val id =  ID

    companion object {
        const val ID = 0
    }
}
